src/main/java/com/google/devtools/build/lib/actions/ActionCacheChecker.java

```diff
@@ -435,6 +441,100 @@ public void updateActionCache(
+    }
+
+    for (Artifact output : action.getOutputs()) {
+      if (output.isConstantMetadata()) {
+        continue;
+      }
+
+      if (output.isTreeArtifact()) {
+        SpecialArtifact artifact = (SpecialArtifact) output;
+        TreeArtifactValue metadata = actionCache.getTreeMetadata(artifact);
+        if (metadata != null) {
+          // If tree files metadata are entirely remote, just use this value.
+          if (metadata.isEntirelyRemote()) {
+            outputStore.putTreeArtifactData(artifact, metadata);
+          } else {
+            // Load remote tree files metadata into cache. Metadata for local tree files should be
+            // reconstructed from local file system.
+            for (Map.Entry<TreeFileArtifact, FileArtifactValue> childEntry :
+                metadata.getChildValues().entrySet()) {
+              TreeFileArtifact child = childEntry.getKey();
+              FileArtifactValue childMetadata = childEntry.getValue();
+              if (childMetadata != MISSING_FILE_MARKER && childMetadata.isRemote()) {
+                outputStore.putTreeFileArtifactData(child, childMetadata);
+              }
+            }
+          }
+        }
+      } else {
+        FileArtifactValue metadata = actionCache.getFileMetadata(output);
+        // Load remote file metadata from cache. Metadata for local files should be reconstructed
+        // from local file system
+        if (metadata != null && metadata.isRemote()) {
+          outputStore.putArtifactData(output, metadata);
+        }
+      }
+    }
+  }
+
+  void updateMetadata(Action action, MetadataHandler metadataHandler) throws IOException {
+    for (Artifact output : action.getOutputs()) {
+      if (!metadataHandler.artifactOmitted(output)) {
+        // Update metadata in the action cache
+        if (output.isTreeArtifact()) {
+          SpecialArtifact artifact = (SpecialArtifact) output;
+          TreeArtifactValue treeArtifactValue = metadataHandler.getTreeArtifactValue(artifact);
+          if (treeArtifactValue.isEntirelyRemote()) {
+            actionCache.putTreeMetadata(artifact, treeArtifactValue);
+          } else {
+            boolean hasRemote = false;
+            for (Map.Entry<TreeFileArtifact, FileArtifactValue> childEntry :
+                treeArtifactValue.getChildValues().entrySet()) {
+              hasRemote |= childEntry.getValue().isRemote();
+            }
+
+            if (hasRemote) {
+              TreeArtifactValue.Builder builder = TreeArtifactValue.newBuilder(artifact);
+              for (Map.Entry<TreeFileArtifact, FileArtifactValue> childEntry :
+                  treeArtifactValue.getChildValues().entrySet()) {
+                TreeFileArtifact child = childEntry.getKey();
+                FileArtifactValue childMetadata = childEntry.getValue();
+                if (!childMetadata.isRemote()) {
+                  // Only save remote tree file metadata
+                  childMetadata = MISSING_FILE_MARKER;
```

Comment:
I wonder if we should have a different marker or at least add a comment that this differs from what that represents -- this is metadata for a file which does not exists at all. In here, we are saying it exists, but locally.